pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -185,6 +185,40 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 	return builder
 }
 
+func buildNameToServiceMapForHTTPRoutes(node *model.Proxy, push *model.PushContext,
+	virtualService config.Config) map[host.Name]*model.Service {
+
+	vs := virtualService.Spec.(*networking.VirtualService)
+	nameToServiceMap := map[host.Name]*model.Service{}
+
+	addService := func(hostname host.Name) {
+		if _, exist := nameToServiceMap[hostname]; exist {
+			return
+		}
+		// First, we obtain the service which has the same namespace as virtualService
+		if s, exist := push.ServiceIndex.HostnameAndNamespace[hostname][virtualService.Namespace]; exist {
+			nameToServiceMap[hostname] = s
+		} else {
+			// Didn't find match for our namespace, pick the service which is visible to the ConfigNamespace of node.
+			nameToServiceMap[hostname] = push.ServiceForHostname(node, hostname)
+		}
+	}
+
+	for _, httpRoute := range vs.Http {
```

Comment:
This is making the assumption that users of nameToServiceMap *only* lookup hosts that are in mirror/route.destination. But it seems not the case? https://github.com/istio/istio/blob/68ebdc1f1eb9fe832e6ddc26962c0630f9275de1/pilot/pkg/networking/core/v1alpha3/route/route.go#L163

Reply:
The case you mentioned above is that RDS generates `virtualhosts` based on `rule.Hosts` in sidecar scene. But in my case, we are building `route.Route` based on `networking.HTTPRoute` of VirtualService in gateway scene. So in the following code logic, we only need the hostname information in `networking.HTTPRoute`, which are mirror/route.destination.

Reply:
Good point - I got mixed up. thanks